compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java

```diff
@@ -1373,8 +1375,20 @@ private boolean validateErrorVariable(BLangErrorVariable errorVariable, BErrorTy
         }
 
         if (isRestDetailBindingAvailable(errorVariable)) {
+            // Type of rest pattern is a map type where constraint type is,
+            // union of keys whose values are not matched in error binding/match pattern.
             BTypeSymbol typeSymbol = createTypeSymbol(SymTag.TYPE);
-            BMapType restType = new BMapType(TypeTags.MAP, recordType.restFieldType, typeSymbol);
+            BUnionType restFieldType = BUnionType.create(null, recordType.restFieldType);
+            for (Map.Entry<String, BField> entry : fieldMap.entrySet()) {
+                if (!matchedDetailFields.contains(entry.getKey())) {
+                    BType type = entry.getValue().getType();
+                    if (!type.equals(symTable.nilType)) {
+                        restFieldType.add(type);
+                    }
+                }
+            }
+
+            BMapType restType = new BMapType(TypeTags.MAP, restFieldType, typeSymbol);
```

Comment:
Shouldn't we first check if the union has more than one member? If it is just one type we don't need a union?

Reply:
It feels like a premature optimization, do you see a downside of single-member union here?

Reply:
IINM, we've done this across such similar union type creations in the compiler.

IMO it would be cleaner to have a union only when there is > 1 type, this could potentially be helpful in subsequent type-checking/analysis when accessing elements from the map too (same type checks would pass and we wouldn't have to work with union types)?

Reply:
I didn't know that it's a common practice to lift out single union members. Will change as suggested.